File: backend/app/agent/tools/executor.py
"""ToolExecutor — dispatches LLM tool calls to their implementations."""

from __future__ import annotations

import json
from typing import Any

from sqlalchemy import or_, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.agent.game_data.dice_engine import CombatantInput, calculate_f2f
from app.agent.game_data.fireteam_logic import get_fireteam_bonuses
from app.agent.game_data.loader import GameDataLoader
from app.models.faction import Faction
from app.models.fireteam import FireteamChart
from app.models.unit import Loadout, Unit
from app.routers.units import (
    _build_loadout_response,
    _build_profile_response,
    _get_catalogs,
)


class ToolExecutor:
    def __init__(self, db: AsyncSession, loader: GameDataLoader) -> None:
        self._db = db
        self._loader = loader
        self._handlers = {
            "search_units": self._search_units,
            "get_unit_profile": self._get_unit_profile,
            "get_faction_context": self._get_faction_context,
            "validate_fireteam": self._validate_fireteam,
            "analyze_matchup": self._analyze_matchup,
            "analyze_classifieds": self._analyze_classifieds,
        }

    async def execute(self, tool_name: str, tool_input: dict[str, Any]) -> str:
        handler = self._handlers.get(tool_name)
        if not handler:
            return json.dumps({"error": f"Unknown tool: {tool_name}"})
        try:
            result = await handler(**tool_input)
            return json.dumps(result)
        except Exception as e:
            return json.dumps({"error": str(e)})

    async def _search_units(
        self,
        query: str | None = None,
        faction_id: int | None = None,
        min_points: int | None = None,
        max_points: int | None = None,
        limit: int = 20,
    ) -> dict[str, Any]:
        q = (
            select(Unit)
            .options(selectinload(Unit.factions), selectinload(Unit.loadouts))
            .distinct()
        )

        if query:
            pattern = f"%{query}%"
            q = q.where(or_(Unit.name.ilike(pattern), Unit.isc.ilike(pattern)))

        if faction_id is not None:
            q = q.join(Unit.factions).where(Faction.id == faction_id)

        if min_points is not None or max_points is not None:
            q = q.join(Unit.loadouts)
            if min_points is not None:
                q = q.where(Loadout.points >= min_points)
            if max_points is not None:
                q = q.where(Loadout.points <= max_points)

        q = q.order_by(Unit.name).limit(min(limit, 30))
        result = await self._db.execute(q)
        units = result.scalars().unique().all()

        return {
            "count": len(units),
            "units": [
                {
                    "name": u.name,
                    "isc": u.isc,
                    "slug": u.slug,
                    "factions": [f.slug for f in u.factions],
                    "points_min": min((lo.points for lo in u.loadouts), default=0),
                    "points_max": max((lo.points for lo in u.loadouts), default=0),
                }
                for u in units
            ],
        }

    async def _get_unit_profile(self, slug: str) -> dict[str, Any]:
        result = await self._db.execute(
            select(Unit)
            .options(
                selectinload(Unit.factions),
                selectinload(Unit.profiles),
                selectinload(Unit.loadouts),
            )
            .where(Unit.slug == slug)
        )
        unit = result.scalar_one_or_none()
        if not unit:
            return {"error": f"Unit '{slug}' not found"}

        weapon_map, skill_map, equip_map = await _get_catalogs(self._db)

        profiles = [
            _build_profile_response(p, weapon_map, skill_map, equip_map)
            for p in sorted(unit.profiles, key=lambda p: (p.profile_group_id, p.id))
        ]
        loadouts = [
            _build_loadout_response(lo, weapon_map, skill_map, equip_map)
            for lo in sorted(
                unit.loadouts[:5],
                key=lambda lo: (lo.profile_group_id, lo.option_id),
            )
        ]

        return {
            "name": unit.name,
            "isc": unit.isc,
            "slug": unit.slug,
            "factions": [f.slug for f in unit.factions],
            "profiles": [
                {
                    "name": p.name,
                    "mov": p.mov,
                    "cc": p.cc,
                    "bs": p.bs,
                    "ph": p.ph,
                    "wip": p.wip,
                    "arm": p.arm,
                    "bts": p.bts,
                    "wounds": p.wounds,
                    "silhouette": p.silhouette,
                    "unit_type": p.unit_type,
                    "skills": [
                        {"name": s.name, "modifiers": s.extra_display} for s in p.skills
                    ],
                    "equipment": [
                        {"name": e.name, "modifiers": e.extra_display}
                        for e in p.equipment
                    ],
                    "weapons": [
                        {"name": w.name, "modifiers": w.extra_display}
                        for w in p.weapons
                    ],
                }
                for p in profiles
            ],
            "loadouts": [
                {
                    "name": lo.name,
                    "points": lo.points,
                    "swc": lo.swc,
                    "weapons": [
                        {"name": w.name, "modifiers": w.extra_display}
                        for w in lo.weapons
                    ],
                    "skills": [
                        {"name": s.name, "modifiers": s.extra_display}
                        for s in lo.skills
                    ],
                    "equipment": [
                        {"name": e.name, "modifiers": e.extra_display}
                        for e in lo.equipment
                    ],
                }
                for lo in loadouts
            ],
        }

    async def _get_faction_context(
        self,
        faction_id: int,
        include_roster: bool = True,
        include_fireteams: bool = True,
    ) -> dict[str, Any]:
        faction_result = await self._db.execute(
            select(Faction).where(Faction.id == faction_id)
        )
        faction = faction_result.scalar_one_or_none()
        if not faction:
            return {"error": f"Faction {faction_id} not found"}

        out: dict[str, Any] = {
            "id": faction.id,
            "name": faction.name,
            "slug": faction.slug,
        }

        if include_fireteams:
            ft_result = await self._db.execute(
                select(FireteamChart).where(FireteamChart.faction_id == faction_id)
            )
            ft_chart = ft_result.scalar_one_or_none()
            if ft_chart:
                teams = ft_chart.chart_json.get("teams", [])
                out["fireteams"] = [
                    {
                        "name": t.get("name"),
                        "type": t.get("type", []),
                        "slots": [
                            {
                                "name": u.get("name"),
                                "min": u.get("min", 0),
                                "max": u.get("max", 1),
                            }
                            for u in t.get("units", [])
                        ],
                    }
                    for t in teams
                ]

        if include_roster:
            q = (
                select(Unit)
                .options(selectinload(Unit.factions), selectinload(Unit.loadouts))
                .join(Unit.factions)
                .where(Faction.id == faction_id)
                .order_by(Unit.name)
                .limit(30)
            )
            units_result = await self._db.execute(q)
            units = units_result.scalars().unique().all()
            out["roster"] = [
                {
                    "name": u.name,
                    "isc": u.isc,
                    "slug": u.slug,
                    "points_min": min((lo.points for lo in u.loadouts), default=0),
                    "points_max": max((lo.points for lo in u.loadouts), default=0),
                }
                for u in units
            ]
            out["roster_truncated"] = len(units) == 30

        return out

    async def _validate_fireteam(
        self,
        faction_id: int,
        team_name: str,
        member_names: list[str],
    ) -> dict[str, Any]:
        ft_result = await self._db.execute(
            select(FireteamChart).where(FireteamChart.faction_id == faction_id)
        )
        ft_chart = ft_result.scalar_one_or_none()
        if not ft_chart:
            return {"error": f"No fireteam chart found for faction {faction_id}"}

        teams = ft_chart.chart_json.get("teams", [])
        team = next(
            (t for t in teams if t.get("name", "").lower() == team_name.lower()),
            None,
        )
        if not team:
            available = [t.get("name") for t in teams]
            return {
                "error": f"Team '{team_name}' not found",
                "available_teams": available,
            }

        members = [{"name": n} for n in member_names]
        bonuses = get_fireteam_bonuses(team, members)

        active_bonuses = [b for b in bonuses if b.is_active]
        return {
            "team": team_name,
            "members": member_names,
            "is_valid": len(active_bonuses) > 0,
            "bonuses": [
                {"level": b.level, "description": b.description, "active": b.is_active}
                for b in bonuses
            ],
            "active_level": max((b.level for b in active_bonuses), default=0),
        }

    async def _analyze_matchup(
        self,
        active_sv: int,
        active_burst: int,
        active_damage: int,
        reactive_sv: int,
        reactive_burst: int,
        reactive_damage: int,
        target_arm: int,
        active_arm: int,
        active_ammo: str = "NORMAL",
        reactive_ammo: str = "NORMAL",
    ) -> dict[str, Any]:
        active = CombatantInput(
            sv=active_sv,
            burst=active_burst,
            damage=active_damage,
            ammo=active_ammo,
            arm=active_arm,
        )
        reactive = CombatantInput(
            sv=reactive_sv,
            burst=reactive_burst,
            damage=reactive_damage,
            ammo=reactive_ammo,
            arm=target_arm,
        )
        result = calculate_f2f(active, reactive)

        return {
            "active_wins_pct": result.active_wins,
            "reactive_wins_pct": result.reactive_wins,
            "draw_pct": result.draw,
            "expected_wounds_on_reactive": result.expected_active_wounds,
            "expected_wounds_on_active": result.expected_reactive_wounds,
            "wound_distribution_on_reactive": result.wound_dist_active,
            "wound_distribution_on_active": result.wound_dist_reactive,
            "summary": (
                f"Active player wins {result.active_wins}% of the time, "
                f"dealing {result.expected_active_wounds:.2f} expected wounds. "
                f"Reactive wins {result.reactive_wins}%."
            ),
        }

    async def _analyze_classifieds(
        self,
        unit_names: list[str],
    ) -> dict[str, Any]:
        classifieds = self._loader.classifieds
        if not classifieds:
            return {"error": "Classifieds data not available"}

        lower_names = {n.lower() for n in unit_names}

        completable = []
        not_completable = []

        for obj in classifieds:
            designated = [t.lower() for t in obj.get("designatedTroopers", [])]
            if not designated:
                completable.append(
                    {"name": obj["name"], "category": obj.get("category", "")}
                )
                continue

            can_complete = any(
                any(d in n or n in d for n in lower_names) for d in designated
            )

            if can_complete:
                completable.append(
                    {
                        "name": obj["name"],
                        "category": obj.get("category", ""),
                        "requires": obj.get("designatedTroopers", []),
                    }
                )
            else:
                not_completable.append(
                    {
                        "name": obj["name"],
                        "category": obj.get("category", ""),
                        "requires": obj.get("designatedTroopers", []),
                    }
                )

        total = len(classifieds)
        pct = round(len(completable) / total * 100) if total else 0

        return {
            "coverage_pct": pct,
            "completable_count": len(completable),
            "total_count": total,
            "completable": completable,
            "not_completable": not_completable,
            "summary": f"Your list can complete {len(completable)}/{total} classifieds ({pct}%).",
        }

File: backend-ts/src/agent/tools/executor.ts
import { and, asc, eq, ilike, inArray, or, gte, lte, type SQL } from 'drizzle-orm';
import { db } from '../../db/client.js';
import {
    units,
    factions,
    unit_factions,
    profiles,
    loadouts,
    fireteam_charts,
} from '../../db/schema.js';
import { resolveItems } from '../../lib/items.js';
import { getCatalogs } from '../../lib/catalogs.js';
import { calculateF2F, type CombatantInput } from '@shared/dice-engine';
import { getFireteamBonuses } from '@shared/fireteams';
import type { Fireteam } from '@shared/types';
import { GameDataLoader } from '../gameData/loader.js';

const ROSTER_CAP = 30;
const LOADOUT_CAP = 5;

type ToolHandler = (input: Record<string, unknown>) => Promise<unknown>;

export class ToolExecutor {
    private handlers: Record<string, ToolHandler>;

    constructor(private loader: GameDataLoader) {
        this.handlers = {
            search_units: input => this.searchUnits(input),
            get_unit_profile: input => this.getUnitProfile(input),
            get_faction_context: input => this.getFactionContext(input),
            validate_fireteam: input => this.validateFireteam(input),
            analyze_matchup: async input => this.analyzeMatchup(input),
            analyze_classifieds: async input => this.analyzeClassifieds(input),
        };
    }

    async execute(toolName: string, input: Record<string, unknown>): Promise<string> {
        const handler = this.handlers[toolName];
        if (!handler) return JSON.stringify({ error: `Unknown tool: ${toolName}` });
        try {
            const result = await handler(input);
            return JSON.stringify(result);
        } catch (e) {
            return JSON.stringify({ error: (e as Error).message });
        }
    }

    // ---- handlers --------------------------------------------------------

    private async searchUnits(input: Record<string, unknown>) {
        const query = asString(input.query);
        const factionId = asInt(input.faction_id);
        const minPoints = asInt(input.min_points);
        const maxPoints = asInt(input.max_points);
        const limit = Math.min(asInt(input.limit) ?? 20, ROSTER_CAP);

        let unitIds: number[] | null = null;

        if (factionId != null) {
            const rows = await db
                .select({ unit_id: unit_factions.unit_id })
                .from(unit_factions)
                .where(eq(unit_factions.faction_id, factionId));
            unitIds = rows.map(r => r.unit_id);
            if (unitIds.length === 0) return { count: 0, units: [] };
        }

        if (minPoints != null || maxPoints != null) {
            const conds: SQL[] = [];
            if (minPoints != null) conds.push(gte(loadouts.points, minPoints));
            if (maxPoints != null) conds.push(lte(loadouts.points, maxPoints));
            const rows = await db
                .selectDistinct({ unit_id: loadouts.unit_id })
                .from(loadouts)
                .where(and(...conds));
            const ids = rows.map(r => r.unit_id);
            if (ids.length === 0) return { count: 0, units: [] };
            unitIds = unitIds ? unitIds.filter(id => ids.includes(id)) : ids;
            if (unitIds.length === 0) return { count: 0, units: [] };
        }

        const conditions: SQL[] = [];
        if (query) {
            const pattern = `%${query}%`;
            conditions.push(or(ilike(units.name, pattern), ilike(units.isc, pattern))!);
        }
        if (unitIds) conditions.push(inArray(units.id, unitIds));

        const baseRows = await db
            .selectDistinct({
                id: units.id,
                isc: units.isc,
                name: units.name,
                slug: units.slug,
            })
            .from(units)
            .where(conditions.length ? and(...conditions) : undefined)
            .orderBy(asc(units.name))
            .limit(limit);

        if (baseRows.length === 0) return { count: 0, units: [] };
        const ids = baseRows.map(u => u.id);

        const [factionRows, loadoutRows] = await Promise.all([
            db
                .select({ unit_id: unit_factions.unit_id, slug: factions.slug })
                .from(unit_factions)
                .innerJoin(factions, eq(factions.id, unit_factions.faction_id))
                .where(inArray(unit_factions.unit_id, ids)),
            db
                .select({ unit_id: loadouts.unit_id, points: loadouts.points })
                .from(loadouts)
                .where(inArray(loadouts.unit_id, ids)),
        ]);

        const factionsByUnit = new Map<number, string[]>();
        for (const r of factionRows) {
            const arr = factionsByUnit.get(r.unit_id) ?? [];
            arr.push(r.slug);
            factionsByUnit.set(r.unit_id, arr);
        }
        const ptsByUnit = new Map<number, number[]>();
        for (const r of loadoutRows) {
            const arr = ptsByUnit.get(r.unit_id) ?? [];
            arr.push(r.points);
            ptsByUnit.set(r.unit_id, arr);
        }

        return {
            count: baseRows.length,
            units: baseRows.map(u => {
                const pts = ptsByUnit.get(u.id) ?? [];
                return {
                    name: u.name,
                    isc: u.isc,
                    slug: u.slug,
                    factions: factionsByUnit.get(u.id) ?? [],
                    points_min: pts.length ? Math.min(...pts) : 0,
                    points_max: pts.length ? Math.max(...pts) : 0,
                };
            }),
        };
    }

    private async getUnitProfile(input: Record<string, unknown>) {
        const slug = asString(input.slug);
        if (!slug) return { error: 'slug is required' };

        const [unit] = await db.select().from(units).where(eq(units.slug, slug)).limit(1);
        if (!unit) return { error: `Unit '${slug}' not found` };

        const [factionRows, profileRows, loadoutRows, catalogs] = await Promise.all([
            db
                .select({ slug: factions.slug })
                .from(unit_factions)
                .innerJoin(factions, eq(factions.id, unit_factions.faction_id))
                .where(eq(unit_factions.unit_id, unit.id)),
            db.select().from(profiles).where(eq(profiles.unit_id, unit.id)),
            db.select().from(loadouts).where(eq(loadouts.unit_id, unit.id)),
            getCatalogs(),
        ]);

        const sortedProfiles = [...profileRows].sort(
            (a, b) => a.profile_group_id - b.profile_group_id || a.id - b.id,
        );
        const sortedLoadouts = [...loadoutRows]
            .sort((a, b) => a.profile_group_id - b.profile_group_id || a.option_id - b.option_id)
            .slice(0, LOADOUT_CAP);

        return {
            name: unit.name,
            isc: unit.isc,
            slug: unit.slug,
            factions: factionRows.map(f => f.slug),
            profiles: sortedProfiles.map(p => ({
                name: p.name,
                mov: `${p.mov_1}-${p.mov_2}`,
                cc: p.cc,
                bs: p.bs,
                ph: p.ph,
                wip: p.wip,
                arm: p.arm,
                bts: p.bts,
                wounds: p.wounds,
                silhouette: p.silhouette,
                unit_type: p.unit_type,
                skills: resolveItems(p.skills_json, catalogs.skills).map(i => ({
                    name: i.name,
                    modifiers: i.extra_display,
                })),
                equipment: resolveItems(p.equipment_json, catalogs.equipment).map(i => ({
                    name: i.name,
                    modifiers: i.extra_display,
                })),
                weapons: resolveItems(p.weapons_json, catalogs.weapons).map(i => ({
                    name: i.name,
                    modifiers: i.extra_display,
                })),
            })),
            loadouts: sortedLoadouts.map(lo => ({
                name: lo.name,
                points: lo.points,
                swc: lo.swc,
                weapons: resolveItems(lo.weapons_json, catalogs.weapons).map(i => ({
                    name: i.name,
                    modifiers: i.extra_display,
                })),
                skills: resolveItems(lo.skills_json, catalogs.skills).map(i => ({
                    name: i.name,
                    modifiers: i.extra_display,
                })),
                equipment: resolveItems(lo.equipment_json, catalogs.equipment).map(i => ({
                    name: i.name,
                    modifiers: i.extra_display,
                })),
            })),
        };
    }

    private async getFactionContext(input: Record<string, unknown>) {
        const factionId = asInt(input.faction_id);
        if (factionId == null) return { error: 'faction_id is required' };
        const includeRoster = (input.include_roster ?? true) !== false;
        const includeFireteams = (input.include_fireteams ?? true) !== false;

        const [faction] = await db.select().from(factions).where(eq(factions.id, factionId)).limit(1);
        if (!faction) return { error: `Faction ${factionId} not found` };

        const out: Record<string, unknown> = {
            id: faction.id,
            name: faction.name,
            slug: faction.slug,
        };

        if (includeFireteams) {
            const [chart] = await db
                .select()
                .from(fireteam_charts)
                .where(eq(fireteam_charts.faction_id, factionId))
                .limit(1);
            if (chart) {
                const teams = (chart.chart_json as { teams?: Fireteam[] })?.teams ?? [];
                out.fireteams = teams.map(t => ({
                    name: t.name,
                    type: t.type ?? [],
                    slots: (t.units ?? []).map(u => ({
                        name: u.name,
                        min: u.min ?? 0,
                        max: u.max ?? 1,
                    })),
                }));
            }
        }

        if (includeRoster) {
            const rosterIds = await db
                .select({ unit_id: unit_factions.unit_id })
                .from(unit_factions)
                .where(eq(unit_factions.faction_id, factionId));
            const ids = rosterIds.map(r => r.unit_id);
            if (ids.length > 0) {
                const baseRows = await db
                    .select({
                        id: units.id,
                        isc: units.isc,
                        name: units.name,
                        slug: units.slug,
                    })
                    .from(units)
                    .where(inArray(units.id, ids))
                    .orderBy(asc(units.name))
                    .limit(ROSTER_CAP);

                const ptsRows = await db
                    .select({ unit_id: loadouts.unit_id, points: loadouts.points })
                    .from(loadouts)
                    .where(inArray(loadouts.unit_id, baseRows.map(b => b.id)));
                const ptsByUnit = new Map<number, number[]>();
                for (const r of ptsRows) {
                    const arr = ptsByUnit.get(r.unit_id) ?? [];
                    arr.push(r.points);
                    ptsByUnit.set(r.unit_id, arr);
                }

                out.roster = baseRows.map(u => {
                    const pts = ptsByUnit.get(u.id) ?? [];
                    return {
                        name: u.name,
                        isc: u.isc,
                        slug: u.slug,
                        points_min: pts.length ? Math.min(...pts) : 0,
                        points_max: pts.length ? Math.max(...pts) : 0,
                    };
                });
                out.roster_truncated = baseRows.length === ROSTER_CAP;
            } else {
                out.roster = [];
                out.roster_truncated = false;
            }
        }

        return out;
    }

    private async validateFireteam(input: Record<string, unknown>) {
        const factionId = asInt(input.faction_id);
        const teamName = asString(input.team_name);
        const memberNames = asStringArray(input.member_names);
        if (factionId == null || !teamName || memberNames.length === 0) {
            return { error: 'faction_id, team_name, and member_names are required' };
        }

        const [chart] = await db
            .select()
            .from(fireteam_charts)
            .where(eq(fireteam_charts.faction_id, factionId))
            .limit(1);
        if (!chart) return { error: `No fireteam chart found for faction ${factionId}` };

        const teams: Fireteam[] = (chart.chart_json as { teams?: Fireteam[] })?.teams ?? [];
        const team = teams.find(t => t.name?.toLowerCase() === teamName.toLowerCase());
        if (!team) {
            return {
                error: `Team '${teamName}' not found`,
                available_teams: teams.map(t => t.name),
            };
        }

        const members = memberNames.map(n => ({ name: n }));
        const bonuses = getFireteamBonuses(team, members);
        const activeBonuses = bonuses.filter(b => b.isActive);

        return {
            team: teamName,
            members: memberNames,
            is_valid: activeBonuses.length > 0,
            bonuses: bonuses.map(b => ({
                level: b.level,
                description: b.description,
                active: b.isActive,
            })),
            active_level: activeBonuses.length
                ? Math.max(...activeBonuses.map(b => b.level))
                : 0,
        };
    }

    private analyzeMatchup(input: Record<string, unknown>) {
        const active: CombatantInput = {
            sv: asInt(input.active_sv) ?? 0,
            burst: asInt(input.active_burst) ?? 0,
            damage: asInt(input.active_damage) ?? 0,
            ammo: asString(input.active_ammo) ?? 'NORMAL',
            arm: asInt(input.active_arm) ?? 0,
            bts: 0,
        };
        const reactive: CombatantInput = {
            sv: asInt(input.reactive_sv) ?? 0,
            burst: asInt(input.reactive_burst) ?? 0,
            damage: asInt(input.reactive_damage) ?? 0,
            ammo: asString(input.reactive_ammo) ?? 'NORMAL',
            arm: asInt(input.target_arm) ?? 0,
            bts: 0,
        };
        const result = calculateF2F(active, reactive);

        return {
            active_wins_pct: result.activeWins,
            reactive_wins_pct: result.reactiveWins,
            draw_pct: result.draw,
            expected_wounds_on_reactive: result.expectedActiveWounds,
            expected_wounds_on_active: result.expectedReactiveWounds,
            wound_distribution_on_reactive: result.woundDistActive,
            wound_distribution_on_active: result.woundDistReactive,
            summary: `Active player wins ${result.activeWins}% of the time, dealing ${result.expectedActiveWounds.toFixed(2)} expected wounds. Reactive wins ${result.reactiveWins}%.`,
        };
    }

    private analyzeClassifieds(input: Record<string, unknown>) {
        const unitNames = asStringArray(input.unit_names);
        const classifieds = this.loader.classifieds;
        if (!classifieds || classifieds.length === 0) {
            return { error: 'Classifieds data not available' };
        }

        const lowerNames = unitNames.map(n => n.toLowerCase());

        const completable: Array<Record<string, unknown>> = [];
        const notCompletable: Array<Record<string, unknown>> = [];

        for (const obj of classifieds) {
            const designated = (obj.designatedTroopers ?? []).map(t => t.toLowerCase());
            if (designated.length === 0) {
                completable.push({ name: obj.name, category: obj.category ?? '' });
                continue;
            }

            const canComplete = designated.some(d =>
                lowerNames.some(n => d.includes(n) || n.includes(d)),
            );
            const entry = {
                name: obj.name,
                category: obj.category ?? '',
                requires: obj.designatedTroopers ?? [],
            };
            if (canComplete) completable.push(entry);
            else notCompletable.push(entry);
        }

        const total = classifieds.length;
        const pct = total ? Math.round((completable.length / total) * 100) : 0;

        return {
            coverage_pct: pct,
            completable_count: completable.length,
            total_count: total,
            completable,
            not_completable: notCompletable,
            summary: `Your list can complete ${completable.length}/${total} classifieds (${pct}%).`,
        };
    }
}

// ---- helpers ------------------------------------------------------------

function asString(v: unknown): string | undefined {
    return typeof v === 'string' && v.length > 0 ? v : undefined;
}
function asInt(v: unknown): number | null {
    if (typeof v === 'number' && Number.isFinite(v)) return Math.trunc(v);
    if (typeof v === 'string') {
        const n = parseInt(v, 10);
        return Number.isFinite(n) ? n : null;
    }
    return null;
}
function asStringArray(v: unknown): string[] {
    return Array.isArray(v) ? v.filter((x): x is string => typeof x === 'string') : [];
}
